crates/objc2/src/runtime/nsobject.rs

```diff
@@ -155,6 +163,36 @@ pub unsafe trait NSObjectProtocol {
     // For example, something may have a return type of `NSString`, while
     // behind the scenes they really return `NSMutableString` and expect it to
     // not be modified.
```

Comment:
Read the comment here; casting e.g. `Id<NSString>` to `Id<NSMutableString>`, even if that is the actual class, is not allowed!

Especially so because `Id<NSString>` is clonable, regardless of what the _actual_ inner class is! So using that, we could trivially violate memory safety:
```rust
let string: Id<NSMutableString> = NSMutableString::new();
let string: Id<NSString> = Id::into_super(string);
let clone: Id<NSString> = string.clone();

let string: Id<NSMutableString> = Id::member_cast(string);
let clone: Id<NSMutableString> = Id::member_cast(clone);

let string: &mut NSMutableString = &mut string;
let clone: &mut NSMutableString = &mut clone;
// `string` and `clone` are now mutable aliasing references, which is UB!
```

Reply:
Ah, I see. The key information I was missing was that mutability is kinda "encoded" in class types. That's why I was under the impression that `Id` would be more like `Rc` (no `DerefMut`) but it's slightly different.

Reply:
Hm. But I feel like it should be possible to do this safely. Since we have knowledge of the complete "inheritance chain" and can only allow casts that don't skip over any `Mutable` classes 🤔.

Reply:
I omitted this in the latest push. You can already kinda do this with `obj.downcast::<...>().unwrap().retain()`.